### PR TITLE
Num real roots sturm

### PR DESCRIPTION
**Pull request by videlec**
Some prerequisites for real root isolation.

### CODE REVIEW

**Comment by videlec** [comment]
In the last three commits:
 - faster `num_real_roots_sturm` version using prs
 - implement `num_real_roots` with shortcut for degrees 2, 3 and 4.

**Comment by videlec** [comment]
With no extra cost it is actually possible to count separately negative and positive roots... From there, it is easy to implement a function to count real roots in intervals [p1/q1, p2/q2] (by mean of a change of variables).

File: fmpz_poly/num_real_roots.c [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        return 0;
+    if (len == 2)
+        return 1;
+    if (len == 3)
+        return _fmpz_poly_num_real_roots_quadratic(pol, len);
+    if (len <= 5)
+    {
+        int s;
+        fmpz_t disc;
+
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots.c]: disc is not initialized

File: fmpz_poly/num_real_roots.c [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            res = 0;
+    }
+
+    fmpz_clear(a);
+    fmpz_clear(d);
+    return res;
+}
+
+
+slong _fmpz_poly_num_real_roots(fmpz * pol, slong len)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots.c]: pol should be const

File: fmpz_poly/num_real_roots_sturm.c [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    FLINT is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <http://www.gnu.org/licenses/>.
+*/
+
+#include "fmpz_poly.h"
+
+
+void _fmpz_poly_num_real_roots_sturm(slong * n_neg, slong * n_pos, fmpz * pol, slong len)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots_sturm.c]: pol should be const

File: fmpz_poly/num_real_roots_sturm.c [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    fmpz_clear(a);
+    fmpz_clear(b);
+    fmpz_clear(g);
+    fmpz_clear(h);
+
+    _fmpz_vec_clear(W, 2*len - 1);
+}
+
+
+slong fmpz_poly_num_real_roots_sturm(fmpz_poly_t pol)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots_sturm.c]: pol should be const

File: fmpz_poly/test/t-num_real_roots.c [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        slong k1, k2;
+        fmpz_poly_t p;
+
+        fmpz_poly_init(p);
+        /* currently the code of num_real_roots only has a special branch */
+        /* for length <= 5. We only test these cases.                     */
+        do
+        {
+            fmpz_poly_randtest_not_zero(p, state, 1 + n_randint(state, 5), 100);
+        }while (!fmpz_poly_is_squarefree(p));
```
Comment (fredrik-johansson) [inline fmpz_poly/test/t-num_real_roots.c]: no space between "}" and "while"

File: fmpz_poly.h [hunk 6]
```diff
@@ -1216,6 +1219,12 @@ FLINT_DLL void _fmpz_poly_bound_roots(fmpz_t bound, const fmpz * poly, slong len
 
 FLINT_DLL void fmpz_poly_bound_roots(fmpz_t bound, const fmpz_poly_t poly);
 
+FLINT_DLL void _fmpz_poly_num_real_roots_sturm(slong * n_neg, slong * n_pos, fmpz * pol, slong len);
```
Comment (fredrik-johansson) [inline fmpz_poly.h]: pol should be const

File: fmpz_poly.h [hunk 7]
```diff
@@ -1216,6 +1219,12 @@ FLINT_DLL void _fmpz_poly_bound_roots(fmpz_t bound, const fmpz * poly, slong len
...
+
+FLINT_DLL slong fmpz_poly_num_real_roots_sturm(fmpz_poly_t poly);
```
Comment (fredrik-johansson) [inline fmpz_poly.h]: poly should be const

File: fmpz_poly.h [hunk 8]
```diff
@@ -1216,6 +1219,12 @@ FLINT_DLL void _fmpz_poly_bound_roots(fmpz_t bound, const fmpz * poly, slong len
...
+
+FLINT_DLL slong fmpz_poly_num_real_roots(fmpz_poly_t poly);
```
Comment (fredrik-johansson) [inline fmpz_poly.h]: poly should be const

File: fmpz_poly/doc/fmpz_poly.txt [hunk 9]
```diff
@@ -3027,6 +3036,31 @@ void fmpz_poly_bound_roots(fmpz_t bound, const fmpz_poly_t poly)
 
     where the coefficients of the polynomial are $a_0, \ldots, a_n$.
 
+void _fmpz_poly_num_real_roots_sturm(slong * n_neg, slong * n_pos,
+            fmpz * pol, slong len)
```
Comment (fredrik-johansson) [inline fmpz_poly/doc/fmpz_poly.txt]: pol should be const

File: fmpz_poly/doc/fmpz_poly.txt [hunk 10]
```diff
@@ -3027,6 +3036,31 @@ void fmpz_poly_bound_roots(fmpz_t bound, const fmpz_poly_t poly)
...
+    Sets \code{n_neg} and \code{n_pos} to the number of negative and
+    positive roots of the polynomial \code{(pol, len)} using Sturm
+    sequence. The Sturm sequence is computed via subresultant
+    remainders obtained by repeated call to the function
+    _fmpz_poly_pseudo_rem_cohen.
+
+    The polynomial is assumed to be squarefree, of degree larger than 1
+    and with non-zero constant coefficient.
+
+slong fmpz_poly_num_real_roots_sturm(fmpz_poly_t pol)
```
Comment (fredrik-johansson) [inline fmpz_poly/doc/fmpz_poly.txt]: pol should be const

File: fmpz_poly/doc/fmpz_poly.txt [hunk 11]
```diff
@@ -3027,6 +3036,31 @@ void fmpz_poly_bound_roots(fmpz_t bound, const fmpz_poly_t poly)
...
+
+    Returns the number of real roots of the squarefree polynomial \code{pol}
+    using Sturm sequence.
+
+    The polynomial is assumed to be squarefree.
+
+slong fmpz_poly_num_real_roots(fmpz_poly_t pol)
```
Comment (fredrik-johansson) [inline fmpz_poly/doc/fmpz_poly.txt]: pol should be const

File: fmpz_poly/num_real_roots.c [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    FLINT is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <http://www.gnu.org/licenses/>.
+*/
+
+#include "fmpz_poly.h"
+
+static __inline__
+slong _fmpz_poly_num_real_roots_quadratic(fmpz * pol, slong len)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots.c]: pol should be const

File: fmpz_poly/num_real_roots.c [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        fmpz_clear(ac);
+        if (s > 0)
+            return 2;
+        else
+            return 0;
+    }
+}
+
+static __inline__
+slong _num_roots_quartic_positive_discriminant(fmpz * p)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots.c]: p should be const

File: fmpz_poly/num_real_roots.c [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    fmpz_init(d);
+
+    /* P = 8ac - 3b^2 */
+    fmpz_mul(d, p + 4, p + 2);
+    fmpz_mul_ui(d, d, 8);
+    fmpz_mul(a, p + 3, p + 3);
+    fmpz_mul_ui(a, a, 3);
+    fmpz_sub(d, d, a);
+
+    if (fmpz_cmp_ui(d, 0) < 0)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots.c]: you can use fmpz_sgn instead

File: fmpz_poly/num_real_roots.c [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        fmpz_mul_ui(a, a, 16);
+        fmpz_sub(d, d, a);
+
+        fmpz_mul(a, p + 3, p + 3);
+        fmpz_mul(a, a, p + 3);
+        fmpz_mul(a, a, p + 3);
+        fmpz_mul_ui(a, a, 3);
+        fmpz_sub(d, d, a);
+
+        if (fmpz_cmp_ui(d, 0) < 0)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots.c]: you can use fmpz_sgn instead

File: fmpz_poly/num_real_roots.c [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+        _fmpz_poly_num_real_roots_sturm(&n_neg, &n_pos, pol + n_zero, len - n_zero);
+        return n_zero + n_neg + n_pos;
+    }
+
+    /* unreachable! */
+    return -1;
+}
+
+slong fmpz_poly_num_real_roots(fmpz_poly_t pol)
```
Comment (fredrik-johansson) [inline fmpz_poly/num_real_roots.c]: pol should be const

**Comment by wbhart** [comment]
Sorry for the delay in reviewing this. Our website was down for a while, meaning the Travis-CI couldn't run. I've restarted the build now to see that everything passes. Fredrik and I are reviewing the code now and Fredrik will post any minor issues we have.

**Comment by fredrik-johansson** [comment]
Looks good to me, apart from the minor issues I've commented on.

**Comment by wbhart** [comment]
Could you check that files have a newline at the end. Some compilers complain miserably about this. I see some files appear to have a blank line when I'm viewing them on GitHub, others don't. I'm not sure if GitHub shows a newline at the end of a file as a blank line at the end or not, though.

File: fmpq_vec/randtest_uniq_sorted.c [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        _fmpq_vec_sort(vec, len);
+        for (i = 0; i < len - 1; i++)
+        {
+            if (fmpq_equal(vec + i, vec + i + 1))
+            {
+                fmpq_randtest(vec + i, state, bits);
+                do_again = 1;
+            }
+        }
+    }while(do_again);
```
Comment (wbhart) [inline fmpq_vec/randtest_uniq_sorted.c]: Need a space before while.

**Comment by videlec** [comment]
done...

**Comment by wbhart** [comment]
Thanks very much. We'll merge this as soon as Travis passes.

**Comment by videlec** [comment]
Thanks!

**Pull request merged**